text/3617-precise-capturing.md

```diff
@@ -0,0 +1,821 @@
+>
+> _ImplTraitTypeOneBound_ :
+> &nbsp;&nbsp; `impl` _UseCaptures_<sup>?</sup> [_TraitBound_][]
+>
+> _UseCaptures_ :\
+> &nbsp;&nbsp; `use` _UseCapturesGenericArgs_
+>
+> _UseCapturesGenericArgs_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; `<` `>` \
+> &nbsp;&nbsp; | `<` \
+> &nbsp;&nbsp; &nbsp;&nbsp; ( _UseCapturesGenericArg_ `,`)<sup>\*</sup> \
+> &nbsp;&nbsp; &nbsp;&nbsp; _UseCapturesGenericArg_ `,`<sup>?</sup> \
+> &nbsp;&nbsp; &nbsp;&nbsp; `>`
+>
+> _UseCapturesGenericArg_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [LIFETIME_OR_LABEL][] \
+> &nbsp;&nbsp; | [IDENTIFIER][]
+
+[IDENTIFIER]: https://doc.rust-lang.org/nightly/reference/identifiers.html
+[LIFETIME_OR_LABEL]: https://doc.rust-lang.org/nightly/reference/tokens.html#lifetimes-and-loop-labels
+[_TraitBound_]: https://doc.rust-lang.org/nightly/reference/trait-bounds.html
+[_TypeParamBounds_]: https://doc.rust-lang.org/nightly/reference/trait-bounds.html
+
+## Reference desugaring
+
+Associated type position `impl Trait` (ATPIT) can also be used, more verbosely, to control capturing of generic parameters in opaque types.  We can use this to describe the semantics of `use<..>`.  If we consider the following code:
+
+```rust
+struct Ty<'u, U, const CU: u8>(&'u (), U);
+
+impl<'u, U, const CU: u8> Ty<'u, U, CU> {
+    pub fn f<'t, T, const CT: u8>(
+        self, x: &'t (), y: T,
+    ) -> impl use<'u, 't, U, T, CU, CT> Sized {
+        (self, x, y, CU, CT)
+    }
+}
+```
+
+Then, using ATPIT, we could desugar this as follows while preserving equivalent semantics:
+
+```rust
+struct Ty<'u, U, const CU: u8>(&'u (), U);
+
+impl<'u, U, const CU: u8> Ty<'u, U, CU> {
+    pub fn f<'t, T, const CT: u8>(
+        self, x: &'t (), y: T,
+    ) -> <() as _0::H>::Opaque<'u, 't, U, T, CU, CT> {
+        <() as _0::H>::f(self, x, y)
+    }
+}
+
+mod _0 {
+    use super::*;
+    pub trait H {
+        type Opaque<'u, 't, U, T, const CU: u8, const CT: u8>;
+        fn f<'u, 't, U, T, const CU: u8, const CT: u8>(
+            s: Ty<'u, U, CU>, x: &'t (), y: T,
+        ) -> Self::Opaque<'u, 't, U, T, CU, CT>;
+    }
+    impl H for () {
+        type Opaque<'u, 't, U, T, const CU: u8, const CT: u8>
+            = impl Sized;
+        #[inline(always)]
+        fn f<'u, 't, U, T, const CU: u8, const CT: u8>(
+            s: Ty<'u, U, CU>, x: &'t (), y: T,
+        ) -> Self::Opaque<'u, 't, U, T, CU, CT> {
+            (s, x, y, CU, CT)
+        }
+    }
+}
+```
+
+## Avoiding capture of higher ranked lifetimes in nested opaques
+
+For implementation reasons, Rust does not yet support higher ranked lifetime bounds on nested opaque types (see [#104288][]).  However, according to the Lifetime Capture Rules 2024, a nested `impl Trait` opaque type *must* capture all generic parameters in scope, including higher ranked ones.  Therefore, in Rust 2024, this code fails to compile:
+
+```rust
+trait Trait { type Ty; }
+impl<F> Trait for F { type Ty = (); }
+
+fn foo() -> impl for<'a> Trait<Ty = impl Sized> {
```

Comment:
```suggestion
fn foo() -> impl for<'a> Trait<'a, Ty = impl Sized> {
```
The lifetime must be constrained by the trait inputs to be used in an associated type. Please change in the later snippet as well.

Reply:
Thanks, I pushed a commit with this and some minor clarifications in this section.

Note that this example is showing a case where we don't want the lifetime captured, so it is indeed not used in the associated type here.  Parameterizing the trait by a lifetime, though, still makes the example stronger since otherwise the `for<..>` binder in the example is really just superfluous and could simply be removed to migrate the code.